internal/kafka/internal/services/kafka.go

```diff
@@ -1319,3 +1307,27 @@ func buildResourceRecordChange(recordName string, clusterIngress string, action
 
 	return resourceRecordChange
 }
+
+func (k *kafkaService) AssignBootstrapServerHost(kafkaRequest *dbapi.KafkaRequest) error {
```

Comment:
Usually a method that changes its input is not a good practice.
Instead of having this method setting the `BootstrapServerHost` attribute of `dbapi.KafkaRequest`, I think would be better to have a `func (k *kafkaService) FindBootstrapServerHost(kafkaRequest dbapi.KafkaRequest) (string, error)` method.
Then the caller would do something like:
```go
	if bootstrapServerHost, err := sevice.FindBootstrapServerHost(kafkaRequest); err != nil {
		// deal with the error
		return nil
	} else {
		kafkaRequest.BootstrapServerHost = bootstrapServerHost
	}
```
What is the advantage of this approach? If a user reads the code, he will know exactly what happened to the `kafkaRequest` object. If the function takes a `pointer` to `kafkaRequest`, then who reads the code will need to inspect the called function has well to search for any side effect.

Reply:
The side effect is intended hence the naming as well `AssignXXX` and passing the pointer. 
The caller, in this case `Prepare..` and the provisiniong reconciler, calls this method passing in a pointer expecting the bootstrap server host to be assigned if no error happened.

Reply:
If the method is to be used by `Prepare..` then I believe it shouldn't be exported.

In both cases I don't see any reason/benefit for not using the proposed pattern: the current one is open to make any change to the received parameters and who reads the code needs to enter the called function to be sure of what is changed.

Reply:
It is not used only in the `Prepare`. It is also used in the reconciler hence the need for it to be exported. 

> In both cases I don't see any reason/benefit for not using the proposed pattern

I agree with having pure functions and methods being side effects safe when they need to. But in this case, I do not think it is needed because this is method is a "setter" as its name indicates.

> the current one is open to make any change to the received parameters and who reads the code needs to enter the called function to be sure of what is changed

The intended behaviour is to assign the bootstrap server host to the kafka just like what the method name says. If someone for "I do not know what reason" change / assigns other things, then the'll need a dedicated method or something else for that.

> who reads the code needs to enter the called function to be sure of what is changed, how it is changed, what the function does when an error occurs, etc. etc.

The same argument can be made with any method. If you are interested in the implementation details of something, you'll need to look it up. But if the method speaks for itself and only does what it does, you care less about the details.

Reply:
> The same argument can be made with any method. If you are interested in the implementation details

That's not accurate: if the method takes a pointer, than it is likely to change the parameter, so I need to see what the method does to know how it changes MY variable. If it takes the param by value, then I don't need to check what the function does to know exactly what will be MY state after the function call.

> The intended behaviour is to assign the bootstrap server host to the kafka just like what the method name says.

Yes. And I don't think it is a good practice. Look at the `append` function for slices. Why do we need to do:
```go
myslice = append(myslice, ....)
```
? `append` is called `append` because it APPENDs values to the slice, so it could simply modify its input.
However, that's not the case: `append` takes a **copy** of the slice and returns a modified slice.

Reply:
> > The same argument can be made with any method. If you are interested in the implementation details
> 
> That's not accurate: if the method takes a pointer, than it is likely to change the parameter, so I need to see what the method does to know how it changes MY variable. If it takes the param by value, then I don't need to check what the function does to know exactly what will be MY state after the function call.

If you are interested in the implementation detail of something, you need to check the details no matter how the method is declared. This is what I was saying. 

> Yes. And I don't think it is a good practice. 

I think it all depends on the context of what the method intended behavior is. A side effect for me, will be if the method said that it is not going to modify the value of the received parameter and ended up modifying it or saying that it is only modifying this one value and ended up modifying another value on top of it. But this is not the case here and if someone changes the implementation to start assigning something else, then they'll need a dedicated method to assign `YYY` (whatever the YYY is) and not this one.

The kafka boostrap server host needs to assigned as part of Kafka preparing step or kafka reassignment step. The two callers of this method call it with the intention of assigning it and nothing else. 

> Look at the append function for slices. 

There are also methods in the core lib that takes in a pointer as a param and assigns value to the received param...

Reply:
> If you are interested in the implementation detail of something

You are missing my point. If the function takes a pointer, I'm **forced** to be interested in the implementation details of the called function to understand how it changes my state

> A side effect for me, will be if the method said that it is not going to modify the value of the received parameter and ended up modifying it or saying that it is only modifying this one value

Again I don't believe again this is correct. The function could contain logic about how to change the value. It could do something like 'if this error happens, assign this, in this other error assign that, otherwise return an error' (this is just an example) or anything else. Without looking at the called function, you can't know what the function does.

Moreover, you will always end looking at the function code because when some error happens, you can't trust the function not changing something else just because it is called 'assignxxx'

> There are also methods in the core lib that takes in a pointer as a param and assigns value to the received param...

I can't find any on top of my mind that are even close to this kind of function.

Anyway: I believe we could discuss for hours without finding an agreement.

This is my conclusion: with the proposed approach there will be the benefit of removing all the side effects and everything we discussed doesn't apply anymore.

OTOH I don't see any benefit in the current approach.

Reply:
> This is my conclusion: with the proposed approach there will be the benefit of removing all the side effects and everything we discussed doesn't apply anymore.

With the proposed approach you are not removing the assignement, you are moving it one level up to the caller.

> OTOH I don't see any benefit in the current approach.

With the current approach, the assignment logic is taken care of by the implementation of the API contract.

> I can't find any on top of my mind that are even close to this kind of function.

The json unmarshalling function for example.

Reply:
> Anyway: I believe we could discuss for hours without finding an agreement.

:-) Yes. I believe there is value in what you are proposing but I feel like there is little of it in the context of this function: or some of the functions like `Prepare()` and `RegisterKafkaJob()` here that also modifies the kafka object. This is my opinion.

I'll leave it to @VanillaSpoon to decide

Reply:
> With the proposed approach you are not removing the assignement, you are moving it one level up to the caller.

With the proposed approach, the assignment is moved to the **owner** of the object. If another function has other side effects, that is another bad practice IMO.

> With the current approach, the assignment logic is taken care of by the implementation of the API contract.

I don't follow you.

> The json unmarshalling function for example.

I believe that method has nothing in common with our method: deserializers must be generic and have no way to know what is the object to be built, so they have to receive the instance to be populated. That's not our case at all.

Anyway:  I really believe there are no benefits at all in using the current approach. However, it is just my opinion: feel free to ignore it.

Reply:
Thanks for both your inputs @ziccardi  and @machi1990 
I took some time to read more into the two options, It was a great learning experience to look at it from both angles. I will leave the approach as it is for now, but I will return to it if there is confusion in the future.